zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Learner.java

```diff
@@ -75,6 +77,7 @@
     protected BufferedOutputStream bufferedOutput;
 
     protected Socket sock;
+    protected boolean sockBeingClosed = false;
```

Comment:
This variable should be volatile as it is accessed from multiple threads.
I think that AtomicBoolean would be better.
So you can enter the real 'close' only once.
If (sockBeingClosed.compareAndSet(false,true))...sock.close()

Reply:
I see it is accessed only from the QuorumPeer thread. Do I miss anything?

Reply:
Yes, I did miss something. Changed to use AtomicBoolean.